src/param.rs

```diff
@@ -470,10 +621,138 @@ impl AudioParamProcessor {
+                        // @note - same strategy as for ramps
+                        matched.cancel_time = Some(event.time);
+                    } else if matched.event_type == AudioParamEventType::SetValueCurveAtTime {
+                        // If 𝐸1 is a setValueCurve with a start time of 𝑡3 and a duration of 𝑑
+                        // If 𝑡𝑐 <= 𝑡3 + 𝑑 :
+                        // Effectively replace this event with a setValueCurve event
+                        // with a start time of 𝑡3 and a new duration of 𝑡𝑐−𝑡3. However,
+                        // this is not a true replacement; this automation MUST take
+                        // care to produce the same output as the original, and not
+                        // one computed using a different duration. (That would cause
+                        // sampling of the value curve in a slightly different way,
+                        // producing different results.)
+                        let start_time = matched.time;
+                        let duration = matched.duration.unwrap();
+
+                        if event.time <= start_time + duration {
+                            matched.cancel_time = Some(event.time);
+                        }
+                    }
+                }
+
+                // [spec] Remove all events with time greater than 𝑡𝑐.
+                self.event_timeline.retain(|queued| {
+                    let mut time = queued.time;
+                    // if the event has a `cancel_time` we use it instead of `time`
+                    if let Some(cancel_time) = queued.cancel_time {
+                        time = cancel_time;
+                    }
+
+                    time <= event.time
+                });
+                continue; // cancel_and_hold events are not inserted timeline
+            }
+
+            // handle SetValueCurveAtTime
+            // @note - These rules argue in favor of having events inserted in
+            // the control thread, let's panic for now
+            //
+            // [spec] If setValueCurveAtTime() is called for time 𝑇 and duration 𝐷
+            // and there are any events having a time strictly greater than 𝑇, but
+            // strictly less than 𝑇+𝐷, then a NotSupportedError exception MUST be thrown.
+            // In other words, it’s not ok to schedule a value curve during a time period
+            // containing other events, but it’s ok to schedule a value curve exactly
+            // at the time of another event.
+            if event.event_type == AudioParamEventType::SetValueCurveAtTime {
+                // check if we don't try to insert at the time of another event
+                let start_time = event.time;
+                let end_time = start_time + event.duration.unwrap();
+
+                for queued in self.event_timeline.iter() {
+                    if queued.time > start_time && queued.time < end_time {
+                        panic!(
```

Comment:
Before releasing a 1.0 version of the lib, we should probably avoid all panics and make these functions return `Result`-like values

Reply:
Yup, I totally agree with that, I just tried to `panic` at places where the spec says an error should be thrown (I missed some of them though) thinking it would be easier to replace them later with a more appropriate error.

I have made a list of todos (spec stuff, optimisations, etc.) while writing the code, I will put all that in the issue #20 (or in a new one if you think it's more appropriate)

Reply:
That sounds great